packages/flutter_tools/lib/src/desktop.dart

```diff
@@ -4,20 +4,31 @@
 
 import 'dart:async';
 
+import 'package:meta/meta.dart';
+
 import 'base/io.dart';
 import 'base/platform.dart';
 import 'base/process_manager.dart';
+import 'commands/daemon.dart';
 import 'convert.dart';
 import 'device.dart';
 import 'version.dart';
 
-// Only launch or display desktop embedding devices if
-// `ENABLE_FLUTTER_DESKTOP` environment variable is set to true.
+@visibleForTesting
+bool debugDisableDesktop = false;
+
+/// Only launch or display desktop embedding devices from the command line
+/// or if `ENABLE_FLUTTER_DESKTOP` environment variable is set to true.
 bool get flutterDesktopEnabled {
-  _flutterDesktopEnabled ??= platform.environment['ENABLE_FLUTTER_DESKTOP']?.toLowerCase() == 'true';
+  if (debugDisableDesktop) {
+    return false;
+  }
+  if (isRunningFromDaemon) {
+    _flutterDesktopEnabled = platform.environment['ENABLE_FLUTTER_DESKTOP']?.toLowerCase() == 'true';
+  }
   return _flutterDesktopEnabled && !FlutterVersion.instance.isStable;
 }
-bool _flutterDesktopEnabled;
+bool _flutterDesktopEnabled = true;
```

Comment:
Since all the logic in the getter is run every time now, it doesn't seem like there's any value in keeping the ivar.